ludwig/backend/ray.py

```diff
@@ -904,61 +919,85 @@ def supports_multiprocessing(self):
+        # normalize NaNs to None
+        column = column.fillna(np.nan).replace([np.nan], [None])
+
         pd_column = self.df_engine.compute(column)
         fnames = pd_column.values.tolist()
-        idxs = pd_column.index.tolist()
 
         # Sample a filename to extract the filesystem info
         sample_fname = fnames[0]
         if isinstance(sample_fname, str):
-            fs, _ = get_fs_and_path(sample_fname)
-
-            read_datasource_fn_kwargs = {
-                "path_and_idxs": list(zip(fnames, idxs)),
-                "filesystem": PyFileSystem(FSSpecHandler(fs)),
-            }
-            if self.df_engine.partitioned and file_size is not None:
-                # Heuristic to determine parallelism: if the average file size is known (in bytes), then we can
-                # extrapolate to determine the total file size. We aim to have ~50MB partitions (5e7 bytes), so we
-                # set parallelism to be the total size / 50MB.
-                total_size = file_size * len(fnames)
-                parallelism = int(total_size / 5e7)
-                # Only set parallelism if it matches or exceeds the Ray default kwarg for parallelism
-                read_datasource_fn_kwargs["parallelism"] = max(RAY_DEFAULT_PARALLELISM, parallelism)
-
-            # The resulting column is named "value", which is a dict with two keys: "idx" and "data".
-            ds = ray.data.read_datasource(BinaryIgnoreNoneTypeDatasource(), **read_datasource_fn_kwargs)
-            df = self.df_engine.from_ray_dataset(ds)
-            # Persist the dataframe to prevent re-reading binary files on each subsequent map_objects call
-            df = self.df_engine.persist(df)
-            df["idx"] = self.df_engine.map_objects(df["value"], lambda row: int(row["idx"]))
-            df["value"] = self.df_engine.map_objects(df["value"], lambda row: row["data"])
-            df = df.rename(columns={"value": column.name})
+            try:
+                import daft
+            except ImportError:
+                raise ImportError(
+                    " daft is not installed. "
+                    "In order to download binary files (like images/audio/..) please run "
+                    "pip install ludwig[distributed]"
+                )
+
+            # Set the runner for executing Daft dataframes to a Ray cluster
+            # Prevent re-initialization errors if the runner is already set
+            # This can happen if there are 2 or more audio/image columns
+            assert ray.is_initialized(), "Ray should be initialized by Ludwig already at application start"
+            daft.context.set_runner_ray(address="auto", noop_if_initialized=True)
+
+            # Convert Dask Series to Dask Dataframe
+            df = column.to_frame(name=column.name)
+            df["idx"] = column.index
+
+            is_dask_df = is_dask_series_or_df(df, self)
+
+            with self.storage.cache.use_credentials():
```

Comment:
Why are we using the cache credentials here, are we reading / writing to the cache?

Reply:
I wasn't sure what else to use to get the right credentials for reading image data from S3 where the user has their own credentials. Would it instead be default credentials?

```
cred_inputs = {
            DEFAULTS: defaults,
            ARTIFACTS: load_creds(artifacts),
            DATASETS: load_creds(datasets),
            CACHE: load_creds(cache),
        }
```

Reply:
Before, we created this FSSpec object and passed it into the RayDataSourceReader

```
fs, _ = get_fs_and_path(sample_fname)

read_datasource_fn_kwargs = {
       "path_and_idxs": list(zip(fnames, idxs)),
       "filesystem": PyFileSystem(FSSpecHandler(fs)),
}
```

Reply:
Maybe we don't need this? But then I don't understand how it was working before..